File: Sem5/Secure Coding/auth_file.c
#include <stdio.h>
#include <string.h>

#define MAX_USERS 5
#define MAX_ATTEMPTS 3
#define USERNAME_LEN 32
#define PASSWORD_LEN 32

typedef struct {
    char username[USERNAME_LEN];
    char password[PASSWORD_LEN];
} User;

int main() {
    User users[MAX_USERS];
    FILE *fp = fopen("password.txt", "r");
    if (!fp) { printf("Could not open password.txt\n"); return 1; }
    int count = 0;
    while (count < MAX_USERS && fscanf(fp, "%31s %31s", users[count].username, users[count].password) == 2) count++;
    fclose(fp);
    char input_username[USERNAME_LEN];
    char input_password[PASSWORD_LEN];
    int attempts = 0, authenticated = 0;
    while (attempts < MAX_ATTEMPTS && !authenticated) {
        printf("Enter username: ");
        if (scanf("%31s", input_username) != 1) break;
        printf("Enter password: ");
        if (scanf("%31s", input_password) != 1) break;
        for (int i = 0; i < count; i++) {
            if (strcmp(users[i].username, input_username) == 0 && strcmp(users[i].password, input_password) == 0) {
                authenticated = 1;
                break;
            }
        }
        if (authenticated) {
            printf("Authentication Successful\n");
        } else {
            attempts++;
            if (attempts < MAX_ATTEMPTS) printf("Authentication failed, try again\n");
        }
    }
    if (!authenticated) printf("Limit exceeded.\n");
    return 0;
}

File: Sem5/Secure Coding/Assignment-1-JS/auth_file.js
const fs = require("fs");
const readline = require("readline-sync");
const filePath = __dirname + '/password.txt';


let data = fs.readFileSync(filePath).toString().trim().split("\n");
let users = {};

for (let i=0; i<data.length; i++){
    let parts = data[i].split(" ");
    users[parts[0]]= parts[i];
}

let maxAttempts = 3;
let attempts = 0;
let auth = false;

while(attempts<maxAttempts && !auth){
    let inputUser = readline.question("Enter Username: ");
    let inputPass = readline.question("Enter Password: ");

    if (users[inputUser] == inputPass){
        console.log("Authentication successful");
        auth = true;
    } 
    else {
        attempts ++;
        if (attempts < maxAttempts){
            console.error("Authentication failed. Please try again.");
            
        }
    }

}

if (!auth && attempts >= maxAttempts){
    console.error("Limit Exceeded");
}